replace ad-hoc context.TODO() with gopts.ctx

armhold opened the pull request:
replace ad-hoc context.TODO() with gopts.ctx, so that cancellation can properly trickle down from cmd_backup & cmd_tag.

This commit fixes the main issue in gh-1434. There are more cmd_* files that need work, but I wanted to give @fd0 a chance to eyeball the pattern of this change before I go further.

armhold committed:
replace ad-hoc context.TODO() with gopts.ctx, so that cancellation
can properly trickle down from cmd_*.

gh-1434

armhold commented:
One potential issue is that the retry code gets kinda verbose on ^c:

```
  Interrupt received, cleaning up
Save(<data/530cf45f5a>) returned error, retrying after 600.230452ms: Copy: context canceled
Save(<data/bcd23ab269>) returned error, retrying after 510.804253ms: Copy: context canceled
Save(<data/e37ea7d9e2>) returned error, retrying after 407.554568ms: Copy: context canceled
Save(<data/75e86b74ce>) returned error, retrying after 538.336984ms: Copy: context canceled
Save(<data/c006f71afb>) returned error, retrying after 340.330567ms: Copy: context canceled
Save(<data/c140597da5>) returned error, retrying after 734.470828ms: Copy: context canceled
Save(<data/e0fa7a31c3>) returned error, retrying after 276.8261ms: Copy: context canceled
Save(<data/2cb60ea073>) returned error, retrying after 749.7146ms: Copy: context canceled
Save(<data/57a46dc104>) returned error, retrying after 708.368778ms: Copy: context canceled
Save(<data/9226a7f8e9>) returned error, retrying after 296.646019ms: Copy: context canceled
Save(<data/276d33f1be>) returned error, retrying after 531.583678ms: Copy: context canceled
Save(<data/fa1a935982>) returned error, retrying after 513.152327ms: Copy: context canceled
Save(<data/e0fa7a31c3>) returned error, retrying after 705.540184ms: Copy: context canceled
Save(<data/9226a7f8e9>) returned error, retrying after 590.725912ms: Copy: context canceled
Save(<data/daa12789cf>) returned error, retrying after 380.688738ms: Copy: context canceled
Save(<data/c006f71afb>) returned error, retrying after 612.777286ms: Copy: context canceled
Save(<data/e37ea7d9e2>) returned error, retrying after 415.888619ms: Copy: context canceled
Save(<data/67421557f1>) returned error, retrying after 312.631071ms: Copy: context canceled
Save(<data/bcd23ab269>) returned error, retrying after 921.963384ms: Copy: context canceled
Save(<data/276d33f1be>) returned error, retrying after 594.247588ms: Copy: context canceled
Save(<data/75e86b74ce>) returned error, retrying after 474.450932ms: Copy: context canceled
Save(<data/530cf45f5a>) returned error, retrying after 399.981093ms: Copy: context canceled
Save(<data/bf2ce785ab>) returned error, retrying after 372.12285ms: Copy: context canceled
Save(<data/fa1a935982>) returned error, retrying after 1.065142243s: Copy: context canceled
Save(<data/daa12789cf>) returned error, retrying after 394.090133ms: Copy: context canceled
```

Wondering if perhaps the retry code should be made aware of ctx and not keep trying when cancelled. Looks like there is a [hook for this](https://github.com/cenkalti/backoff/blob/master/context.go).

fd0 commented:
This is a good first step, thanks! I think we should explicitly test for `context.Canceled` in the backend retry code in `internal/backend/backend_retry.go`, `if errors.Cause(err) == context.Canceled` somewhere.

armhold committed:
make retry code context-aware.

detect cancellation in backend, so that retry code does not keep trying
once user has hit ^c

gh-1434

armhold commented:
>  I think we should explicitly test for context.Canceled in the backend retry code in internal/backend/backend_retry.go

I have implemented this for us to experiment with in a229678, but I have some concerns. Please check https://github.com/armhold/restic/blob/propagate-context/internal/backend/backend_retry.go#L45-L47. Have I understood your intention regarding checking the error cause?

My concern is that if we "throw away" errors in this way, it's possible that the code will continue on in a broken fashion before detecting the error higher up. Like if we're in a be.Save() and the context is canceled, and the data is therefore NOT actually saved, there might be some code higher up that will go on to do some more work expecting that data has been saved when in fact it was not saved.

What do you think?

armhold commented:
Not sure what happened with that integration failure. I can do a clean build and pass all tests (excluding FUSE) on my branch. My branch is lacking several recent commits to master, so perhaps it's a merge issue with newer code.

fd0 commented:
When your branch is merged into master the test fail. There's a new file `restorer_test.go` which has been added in master but is not present in this branch, which also needs fixing.

So you need to rebase your branch on master, fix the test, and then force-push to this PR.

armhold committed:
add context to repo.Flush() call in test code

gh-1434

fd0 commented:
Thanks for sharing your concern, and I think you're right. Please remove the code throwing away the errors. I've had a look and played around with it, and passing the context to the retry logic via https://godoc.org/github.com/cenkalti/backoff#BackOffContext makes sure that retry stops when the context is cancelled. That's what we wanted to achieve :)

armhold committed:
remove check for context.Canceled

gh-1434

armhold committed:
use global context for cat and ls

gh-1434

armhold committed:
use global context for check, debug, dump, find, forget, init, key,
list, mount, tag, unlock commands

gh-1434

armhold commented:
I have gotten all of the TODO()s that I felt comfortable touching. This should solve the lock contention issue, and allow ^c to properly interrupt now.

I also removed the if errors.Cause(err) == context.Canceled as we discussed. FYI, this is what a stack trace will look like without that check:
```
scan [/Users/armhold/work/go/src/github.com/restic/restic]
scanned 1248 directories, 5898 files in 0:00
  signal interrupt received, cleaning up

error while saving data to the repo: context canceled
Close
github.com/restic/restic/internal/backend/b2.(*b2Backend).Save
	src/github.com/restic/restic/internal/backend/b2/b2.go:214
github.com/restic/restic/internal/backend.(*RetryBackend).Save.func1
	src/github.com/restic/restic/internal/backend/backend_retry.go:71
github.com/restic/restic/vendor/github.com/cenkalti/backoff.RetryNotify
	src/github.com/restic/restic/vendor/github.com/cenkalti/backoff/retry.go:37
github.com/restic/restic/internal/backend.(*RetryBackend).retry
	src/github.com/restic/restic/internal/backend/backend_retry.go:37
github.com/restic/restic/internal/backend.(*RetryBackend).Save
	src/github.com/restic/restic/internal/backend/backend_retry.go:65
github.com/restic/restic/internal/repository.(*Repository).savePacker
	src/github.com/restic/restic/internal/repository/packer_manager.go:107
github.com/restic/restic/internal/repository.(*Repository).SaveAndEncrypt
	src/github.com/restic/restic/internal/repository/repository.go:253
github.com/restic/restic/internal/repository.(*Repository).SaveBlob
	src/github.com/restic/restic/internal/repository/repository.go:618
github.com/restic/restic/internal/archiver.(*Archiver).Save
	src/github.com/restic/restic/internal/archiver/archiver.go:106
github.com/restic/restic/internal/archiver.(*Archiver).saveChunk
	src/github.com/restic/restic/internal/archiver/archiver.go:171
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:2337
panic: Close: context canceled

goroutine 1967 [running]:
github.com/restic/restic/internal/archiver.(*Archiver).saveChunk(0xc420454480, 0x1afd760, 0xc4201c2740, 0x3bf4b1, 0x185ca8, 0x9272a9d800000, 0xc427b02000, 0x185ca8, 0x188000, 0xc4205c8000, ...)
	src/github.com/restic/restic/internal/archiver/archiver.go:176 +0x443
created by github.com/restic/restic/internal/archiver.(*Archiver).SaveFile
	src/github.com/restic/restic/internal/archiver/archiver.go:250 +0x3ff
```

This is a little ugly, but IMHO better than carrying on in a bad state. What do you think? 

I think in the future a proper fix to eliminate the stack trace would be to modify saveChunk() (and everything that calls it) to return an error instead of panic(). That would be a substantial design change
though, and not really something I'm prepared to tackle.

fd0 commented:
Thanks for your work, that looks good. The stacktrace not, of course, but I'll tackle that one myself.
